deepliif/data/__init__.py

```diff
@@ -74,11 +77,33 @@ def __init__(self, dataset, batch_size, serial_batches, num_threads, max_dataset
         self.max_dataset_size = max_dataset_size
         self.dataset = dataset
         print("dataset [%s] was created" % type(self.dataset).__name__)
+
+        sampler = None
+        if os.getenv('LOCAL_RANK') is not None or os.getenv('RANK') is not None:
+            sampler = DistributedSampler(self.dataset) if len(gpu_ids) > 0 else None
+
+        # control randomness: https://pytorch.org/docs/stable/notes/randomness.html#dataloader
+        import numpy as np
+        import random
+        def seed_worker(worker_id):
+            worker_seed = torch.initial_seed() % 2**32
+            np.random.seed(worker_seed)
+            random.seed(worker_seed)
+
+        g = torch.Generator()
+        g.manual_seed(0)
+
         self.dataloader = torch.utils.data.DataLoader(
             self.dataset,
+            sampler=sampler,
             batch_size=batch_size,
-            shuffle=not serial_batches,
-            num_workers=int(num_threads))
+            shuffle=not serial_batches if sampler is None else False,
+            num_workers=int(num_threads),
+            worker_init_fn=seed_worker,
+            generator=g
+        )
+
+        self.sampler=sampler            
```

Comment:
This assignment seems to have no effect and could be removed?

Reply:
@ricdodds this is needed because in `cli.py` we would call dataloader.sampler.set_epoch() in each epoch, if DDP is used & shuffle is preferred